scylla-cdc/src/stream_reader.rs

```diff
@@ -134,17 +139,44 @@ impl StreamReader {
             values[begin_index] = CqlValue::Timestamp(window_begin);
             values[end_index] = CqlValue::Timestamp(window_end);
 
-            let mut rows_stream = self
-                .session
-                .execute_iter(query_base.clone(), &values)
-                .await?;
-
-            let schema = CDCRowSchema::new(rows_stream.get_column_specs());
-
-            while let Some(row) = rows_stream.next().await {
-                consumer
-                    .consume_cdc(CDCRow::from_row(row?, &schema))
-                    .await?;
+            let mut sleep_after_timeout = BASIC_TIMEOUT_SLEEP;
+
+            let mut next_state = None;
+            'try_query: loop {
+                let state_clone = next_state.clone();
+                let query_res = self
+                    .session
+                    .execute_paged(&query_base, &values, next_state)
+                    .await;
+                match query_res {
+                    Ok(x) => {
+                        sleep_after_timeout = BASIC_TIMEOUT_SLEEP;
+                        if let Some(rows) = x.rows {
+                            let schema = CDCRowSchema::new(&x.col_specs);
+
+                            for row in rows {
+                                consumer
+                                    .consume_cdc(CDCRow::from_row(row, &schema))
+                                    .await?;
+                            }
+                        }
+                        match x.paging_state {
+                            Some(state) => next_state = Some(state),
+                            None => break 'try_query,
+                        }
+                    },
+                    Err(QueryError::TimeoutError) => {
+                        sleep(time::Duration::from_millis(sleep_after_timeout)).await;
+                        sleep_after_timeout = sleep_after_timeout * TIMEOUT_FACTOR;
+                        if sleep_after_timeout >= TIMEOUT_LIMIT {
+                            println!("WARN: The database keeps returning a timeout error. Trying again...");
+                            sleep_after_timeout = TIMEOUT_LIMIT;
+                        }
+
+                        next_state = state_clone;
+                    },
+                    Err(e) => panic!("The database returned the following error: {:?}", e),
```

Comment:
We shouldn't panic here. This is not an unrecoverable error, we should gracefully stop and propagate the error. Please return it from the function.

Please also use the `Error::context` method to add some information about the source of the error, i.e. that it happened while trying to fetch rows from cdc logs (to differentiate this error from a user error).